pkg/lib/cockpit-components-listing-panel.jsx

```diff
@@ -0,0 +1,160 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2020 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import { Nav, NavItem, NavList, NavVariants } from '@patternfly/react-core';
+import './cockpit-components-listing-panel.scss';
+
+/* tabRenderers optional: list of tab renderers for inline expansion, array of objects with
+ *     - name tab name (has to be unique in the entry, used as react key)
+ *     - renderer react component
+ *     - data render data passed to the tab renderer
+ *     - presence 'always', 'onlyActive', 'loadOnDemand', default: 'loadOnDemand'
+ *         - 'always' once a row is expanded, this tab is always rendered, but invisible if not active
+ *         - 'onlyActive' the tab is only rendered when active
+ *         - 'loadOnDemand' the tab is first rendered when it becomes active, then follows 'always' behavior
+ * listingActions optional: buttons that are presented as actions for the expanded item
+ * simpleBody optional: if set the expansion will just contain this simple body without tabs,
+ *                      this does not work well with tabRenderers.
+ */
+export class ListingPanel extends React.Component {
+    constructor(props) {
+        super(props);
+        const loadedTabs = {};
+        // see if we should preload some tabs
+        let tabPresence;
+        for (let tabIdx = 0; tabIdx < props.tabRenderers.length; tabIdx++) {
+            if ('presence' in props.tabRenderers[tabIdx])
+                tabPresence = props.tabRenderers[tabIdx].presence;
+            else
+                tabPresence = 'default';
+            // the active tab is covered by separate logic
+            if (tabPresence == 'always')
+                loadedTabs[tabIdx] = true;
+        }
+        // ensure the active tab is loaded
+        loadedTabs[props.initiallyActiveTab || 0] = true;
```

Comment:
Here you declare loadedTabs as object but later access it as based on index? Shouldn't loadedTabs be array?

Reply:
That is not new code - it's just code move from pkg/lib/cockpit-components-listing.jsx

It's just an object indexed by numbers, which indeed is confusing.
Let's fix it in another PR, since this PR did not introduce this.